modules/grizzly/src/main/java/org/glassfish/grizzly/threadpool/VirtualThreadExecutorService.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * @author mazhen
+ */
+public class VirtualThreadExecutorService extends AbstractExecutorService implements Thread.UncaughtExceptionHandler {
+
+    private static final Logger logger = Grizzly.logger(VirtualThreadExecutorService.class);
+
+    private final ExecutorService internalExecutorService;
+    private Semaphore poolSemaphore;
+
+    public static VirtualThreadExecutorService createInstance() {
+        return createInstance(ThreadPoolConfig.defaultConfig().setMaxPoolSize(-1).setPoolName("Grizzly-virt-"));
+    }
+
+    public static VirtualThreadExecutorService createInstance(ThreadPoolConfig cfg) {
+        Objects.requireNonNull(cfg);
+        return new VirtualThreadExecutorService(cfg);
+    }
+
+    protected VirtualThreadExecutorService(ThreadPoolConfig cfg) {
+        internalExecutorService = Executors.newThreadPerTaskExecutor(getThreadFactory(cfg));
+        if (cfg.getMaxPoolSize() > 0) {
+            poolSemaphore = new Semaphore(cfg.getMaxPoolSize());
+        } else {
+            poolSemaphore = new Semaphore(Integer.MAX_VALUE);
+        }
+    }
+
+    private ThreadFactory getThreadFactory(ThreadPoolConfig threadPoolConfig) {
+
+        var prefix = threadPoolConfig.getPoolName() + "-";
+
+        // virtual threads factory
+        final ThreadFactory factory = Thread.ofVirtual()
+                .name(prefix, 0L)
+                .uncaughtExceptionHandler(this)
+                .factory();
+
+        return r -> {
+            Thread thread = factory.newThread(r);
+            final ClassLoader initial = threadPoolConfig.getInitialClassLoader();
+            if (initial != null) {
+                thread.setContextClassLoader(initial);
+            }
+            return thread;
+        };
+    }
+
+    @Override
+    public void shutdown() {
+        internalExecutorService.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+        return internalExecutorService.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return internalExecutorService.isShutdown();
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return internalExecutorService.isTerminated();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return internalExecutorService.awaitTermination(timeout, unit);
+    }
+
+    @Override
+    public void execute(Runnable command) {
+        if (poolSemaphore.tryAcquire()) {
```

Comment:
I would also introduce a queue for the tasks that can't be immediately executed. It could be implemented by another semaphore wrapping the poolSemaphore, with the `queueSize + maxPoolSize` permits, like this:

```
Semaphore poolSemaphore = new Semaphore(maxPoolSize, true);
Semaphore queueSemaphore = new Semaphore(queueSize + maxPoolSize, true);

if (queueSemaphore.tryAcquire() {
  poolSemaphore.acquire();
  internalExecutorService.execute(() -> {
                try {
                    command.run();
                } finally {
                    poolSemaphore.release();
                    queueSemaphore.release();
                }
            });
} else {
  throw new RejectedExecutionException("Too Many Concurrent Requests");
}
```

Reply:
Following your latest advice, I've added a second semaphore (`queueSemaphore`) to further refine the task management in `VirtualThreadExecutorService`. 

Your guidance has been instrumental in evolving the functionality and robustness of this service. I'm looking forward to any additional thoughts or feedback you might have!